Add "AIPs by file format" and "AIPs by PUID" reports

tw4l opened the pull request:
Connected to #75 

This PR adds reports that take a file format or PUID as input and provide a sorted list of AIPs containing that format/PUID.

This was intended to show/explore what's possible in terms of format reporting with the data we have from the METS files, so higher-level groupings (e.g. format groups from the FPR) are not included.

Also included is a second commit which simply removes an outdated `TODO` comment on the largest files report, which didn't seem worth the overhead of a separate issue and PR.

## Screenshots

**Reports page**

![image](https://user-images.githubusercontent.com/6758804/96650740-61b3cb00-1301-11eb-8212-32fe2a6513c2.png)

The dropdown lists are populated from new `unique_file_formats` and `unique_puids` properties on the `StorageService` model, and so only contain formats/PUIDs found in the AIPs in the currently selected Storage Service. It's possible these dropdowns might get a bit unwieldy when AIPscan is pointed to a Storage Service with lots of AIPs containing a heterogeneous mix of formats, with an upper bound of being equivalent to the very large dropdown menus in the FPR.

I'm very open to any suggestions for how to make the UI more appealing or user-friendly!

**AIPs by File Format report**

![image](https://user-images.githubusercontent.com/6758804/96650776-6f695080-1301-11eb-9f8c-7853de0fca3f.png)

**AIPs by PUID report**

![image](https://user-images.githubusercontent.com/6758804/96650806-7e500300-1301-11eb-9f8e-a161900b828d.png)

tw4l commented:
So far I've stuck to tabular reports, but these reports might do well in pie chart/scatterplot form as well - maybe something to consider (whether in the scope of this PR or a follow-up).

tw4l commented:
Hey @ross-spencer and @peterVG. I spent most of today starting to implement the start/end date parameters and a pie chart representation of these reports and came across at least one blocker (integration tests) and many questions that I think might take some time to resolve. So in the end I moved that WIP work [to a new branch](https://github.com/artefactual-labs/AIPscan/tree/dev/issue-75-extended) so that I can continue to work on that without it blocking it the tabular reports here, which I think are pretty solid and well-tested.

If you agree that's sensible, this is now ready for re-review.

ross-spencer reviewed (commented):
@tw4l I recognize that this CR might look more critical than it is meant to be. Just arriving back I suspect I've missed some conversations or requirements work here. I am happy to chat more offline and in Slack - and perhaps when we all catch up too? Please review the inline comments and let me know. My summary is such:

* There's a lot of great work here and using a base report is a nice idea. 
* The tests look great and it's good to see those developing.
* There some similarity with existing overview endpoints which I mention inline.
* We should try to avoid literals inline in the Python work and replace them instead with consts at an appropriate scope for both ease of changing and reuse. 
* It seems to me like there is a more potential in this work to collapse either set of functions/reports into the other which I have described inline, i.e. the only difference between the reports seems to be 
   1. how the lookup happens and 
   2. in one report we get the PUID and in the other we don't. 
* We might need to do some more work with format version otherwise we lose granularity in the format lookup in the format (not puid report)
* What are the implications to the user and their understanding of results when we're not distinguishing between original and preservation copy?

I don't think I'm at my peak in terms of writing out these thoughts so if anything is unclear that's on me - as noted, happy to chat more, and I can describe anything in more detail as needed too.

tw4l commented on AIPscan/Reporter/views.py:
```diff
@@ -133,16 +136,30 @@ def view_file(file_id):
+        storage_service = StorageService.query.filter_by(default=True).first()
     all_storage_services = StorageService.query.all()
+    file_formats = storage_service.unique_file_formats
+    # Sort PUIDs "naturally" - i.e. ["fmt/1", "fmt/2", "fmt/10"] rather
+    # than Python's default ["fmt/1", "fmt/10", "fmt/2"].
+    puids = natsorted(storage_service.unique_puids)
```
I see your point about the self-describing name. I added the comment because "natural" is a rather vague term, so I wasn't sure if people would know what it meant without the clarification even if they could glean that `natsorted` means "naturally sorted". I'm fine with separating it into its own helper function if that helps clarify intent. I see what you mean about extensibility as well, though I wonder if that's a bit premature.

On adding this directly into the `render_template` args, I see your point but I think this might be complicated by https://github.com/artefactual-labs/AIPscan/pull/76#discussion_r511143890, namely needing some sort of exception handling for `AttributeError`s when we don't have a Storage Service.

ross-spencer replied:
> I see what you mean about extensibility as well, though I wonder if that's a bit premature.

I am really just focusing on what is clean and self-describing here but in terms of constructive code-review I just want to make sure I share what I do perceive to be benefits. I'm not at all suggesting we will be extending any of this right now.

tw4l replied:
> in terms of constructive code-review I just want to make sure I share what I do perceive to be the benefits if we go the extra step ... I'm just trying to use the comment as a cue to say, okay, this is useful to one of us, how do we improve on that?

Understood, and I appreciate the constructive criticism! I appreciate the nudge to improve on the execution.

tw4l commented on AIPscan/API/namespace_data.py:
```diff
@@ -114,3 +114,43 @@ def get(self, storage_service_id, file_type=None, limit=20):
+        "list_aips_by_format",
+        params={
+            "file_format": {
+                "description": "File format name (must be exact match)",
```
I completely agree that an additional API endpoint that lists the file format names would be very useful here. I think I've been GUI-centric in my thinking because of the immediate priorities, so I didn't think about including that additional API endpoint in the scope of this PR. But I'm happy to revisit that.

On the endpoints, see my comments in the main thread.

tw4l commented on AIPscan/API/namespace_data.py:
```diff
@@ -114,3 +114,43 @@ def get(self, storage_service_id, file_type=None, limit=20):
+        """AIPs containing file format"""
+        file_format = request.args.get("file_format")
+        if not file_format:
+            return {"success": "False", "error": "Must specify a file format."}
```
I will change this to returning an empty list for no match, as I agree that that's much cleaner. As noted above, I had been thinking of the API as "nice to have" rather than a primary goal, hence this rather simple but crude error response.

tw4l replied:
The design intent here was to signal to the caller that their call was invalid because of a missing required parameter. I completely agree that if the parameter is included but no results are found, the best response is to return an empty list. That's what's returned currently if you pass a value for which no results are found, e.g.:

```
{
  "StorageName": "Docker AMSS",
  "Format": "some invalid value",
  "AIPs": []
}
```

Your question about headers is interesting, and a good one that I think warrants some investigation. My `return {"success": "False", "error": "Must specify a file format."}` solution was a bit crude and the concerns you raise around handling outputs are valid ones. I'll plan to do a bit of reading on FlaskX to see if there's a better way to handle missing params.

ross-spencer replied:
Thanks Tessa - yeah, I think the intention is clear, unless I suggested otherwise? I think it's a nice idea too - and really emphasizes your desire for usability. 

I can't seem to find the comment, it might have arrived by email first and then you re-drafted? But I believe you mentioned viewing the API as a sort of nice to have - that's a pretty fair comment at this stage and I might have better remembered that in terms of guidance during review too. If it helps I might leverage that perspective for now and say, let's leave the API fairly minimal in what it's doing and returning and we can then take a more opinionated approach to shaping it a little further down the line. We're not very many iterations down the line yet, and we should each definitely put our heads together about what we want to do as well as avoid here.

tw4l replied:
Oh a second recording, you're totally right and I missed the gist of what you're getting at.

I did do some re-drafting of comments and might have lost that API bit, so I'm glad it came through regardless. Sounds like a good plan to me.

tw4l commented on AIPscan/models.py:
```diff
@@ -46,6 +46,28 @@ def __init__(
+    @property
+    def unique_file_formats(self):
+        formats = (
+            db.session.query(File.file_format.distinct().label("name"))
```
See comments in main thread. This is by design, not omission (a "feature not a bug" :laughing:)

tw4l commented on AIPscan/Data/data.py:
```diff
@@ -292,3 +293,123 @@ def largest_files(storage_service_id, file_type=None, limit=20):
+def _aips_by_format_query(storage_service_id, file_format):
+    """Fetch information on all AIPs with given format from database
+
+    This is separated into its own helper function to aid in testing.
```
Agreed! This is a holdover from the process that I will remove.

tw4l commented on AIPscan/Data/data.py:
```diff
@@ -292,3 +293,123 @@ def largest_files(storage_service_id, file_type=None, limit=20):
+            AIP.transfer_name.label("name"),
+            AIP.uuid.label("uuid"),
+            db.func.count(File.id).label("file_count"),
+            db.func.sum(File.size).label("total_size"),
```
See comment in main thread about reusing the existing endpoints vs introducing new ones.

On the naming of these labels, I see your point and will do some thinking about e.g. using our existing `FIELD_X` constants rather than string literals here (at least outside of the templates, though there may be a way of addressing that).

tw4l replied:
Hey @ross-spencer, coming back to this one a bit - I'm having a hard time replacing this string literals because of how they're only used within the scope of the Data endpoint and accessed as properties of query results.

To illustrate what I mean, we have these labels in `_query_aips_by_file_format_or_puid` (as of https://github.com/artefactual-labs/AIPscan/pull/76/commits/e12649da84d9b80ad882708810758dea5601f734):

```
aips = (
    db.session.query(
        AIP.id.label("id"),
        AIP.transfer_name.label("name"),
        AIP.uuid.label("uuid"),
        db.func.count(File.id).label("file_count"),
        db.func.sum(File.size).label("total_size"),
    )
    ...
```
These labels are only accessed in `_aips_by_file_format_or_puid` here as properties of the query results:

```
...
for result in results:
    aip_info = {}

    aip_info["id"] = result.id
    aip_info[FIELD_AIP_NAME] = result.name
    aip_info[FIELD_UUID] = result.uuid
    aip_info[FIELD_COUNT] = result.file_count
    aip_info[FIELD_SIZE] = result.total_size

    report[FIELD_AIPS].append(aip_info)

return report
```
Since the `FIELD_` constants are being used for the Data endpoint's return value, that's all the API or Reporter blueprints will see - the labels aren't referred to again after this point.

I could make constants for the SQLAlchemy labels, but AFAICT the only place they'd be used is within `_query_aips_by_file_format_or_puid`. Another option would be to remove the labels from the query, in which case we could modify the `_aips_by_file_format_or_puid` code to access data by index - something like this, which I find less readable:

```
...
for result in results:
    aip_info = {}

    aip_info["id"] = result[0]
    aip_info[FIELD_AIP_NAME] = result[1]
    aip_info[FIELD_UUID] = result[2]
    aip_info[FIELD_COUNT] = result[3]
    aip_info[FIELD_SIZE] = result[4]

    report[FIELD_AIPS].append(aip_info)

return report
```

Given all that I'm tempted to leave this as-is. Does that make sense to you?

ross-spencer replied:
I see what you mean Tessa and I appreciate you following the logic through for the demonstration here. I think you are right. 

My only other suggestion is it would be nice to do the following:
```python
    AIP_ID = "id"
    TRANSFER_NAME = "name"
    AIP_UUID = "uuid"
    FILE_COUNT = "file_count"
    FILE_SIZE = "total_size"
    aips = (
        db.session.query(
            AIP.id.label(AIP_ID),
            AIP.transfer_name.label(TRANSFER_NAME),
            AIP.uuid.label(AIP_UUID),
            db.func.count(File.id).label(FILE_COUNT),
            db.func.sum(File.size).label(FILE_SIZE),
        )
        .join(File)
        .join(StorageService)
        .filter(StorageService.id == storage_service_id)
        .group_by(AIP.id)
        .order_by(db.func.count(File.id).desc(), db.func.sum(File.size).desc())
    )
```
This might be colored by my experience of some of the larger code-bases we're working with, because this quite a nice small function - but my usual rationale it's a little cleaner and lends itself better to refactor/extensibility to specify the literals as consts/variables. Any other benefits are minor, but I recognize there is some pragmatism nesting the literals in-line.

tw4l commented on AIPscan/Data/data.py:
```diff
@@ -292,3 +293,123 @@ def largest_files(storage_service_id, file_type=None, limit=20):
+
+    :returns: SQLAlchemy query results
+    """
+    aips = (
+        db.session.query(
+            AIP.id.label("id"),
+            AIP.transfer_name.label("name"),
+            AIP.uuid.label("uuid"),
+            db.func.count(File.id).label("file_count"),
+            db.func.sum(File.size).label("total_size"),
+        )
+        .join(File)
+        .join(StorageService)
+        .filter(StorageService.id == storage_service_id)
+        .filter(File.puid == puid)
+        .group_by(AIP.id)
+        .order_by(db.func.count(File.id).desc(), db.func.sum(File.size).desc())
+    )
+    return aips
```
Point taken and added to list of things to address in follow-up commits

tw4l commented on AIPscan/Data/tests/test_aips_by_puid.py:
```diff
@@ -0,0 +1,57 @@
+
+@pytest.mark.parametrize(
+    "query_results, results_count",
+    [([], 0), (MOCK_QUERY_RESULTS, 4), (MOCK_QUERY_RESULTS[:2], 2)],
```
Excellent suggestions!

tw4l commented on AIPscan/Reporter/report_aips_by_format.py:
```diff
@@ -0,0 +1,26 @@
+    aip_data = data.aips_by_file_format(
+        storage_service_id=storage_service_id, file_format=file_format
+    )
+    storage_service_name = aip_data[data.FIELD_STORAGE_NAME]
```
Agreed

tw4l commented on AIPscan/Reporter/report_aips_by_puid.py:
```diff
@@ -0,0 +1,49 @@
+
+
+def get_format_string_from_puid(puid):
+    """Return file format and version info for a PUID or None
```
See main thread re: intent here. I agree that there's room for sharing code between the reports, but they are intended to communicate different levels of detail about file formats and versions. My intent with this helper was really just to provide a helper so that users looking at a report don't have to go look up what a PUID is externally.

tw4l commented:
Thanks for the feedback, @ross-spencer. You've given me a lot to think about here and I appreciate the time and care you put into these comments.

I'm going to try to start with the "meatiest" issues first and work from there. We do have some opportunities to chat synchronously this week as well, so perhaps we can continue this conversation and add some more nuance then.

> We might need to do some more work with format version otherwise we lose granularity in the format lookup in the format (not puid report)

(also see inline comment re: `StorageService.unique_file_formats` property)

I definitely could have done a better job of articulating the purpose of these reports. In this case, based on conversation in our last sprint meeting and my own interpretation of the requirements, the lost granularity is a "feature not a bug".

We talked in the sprint meeting about wanting reporting on file formats to be possible on several different levels:

1. The individual file format (say all versions of a GIF)
2. The particular file format + version (say GIF 1989a)
3. Format groups (say all raster images; or all flavours of PDF, since they are treated as different formats by PRONOM)

My intention was to have AIPs by File Format meet the need for 1, and AIPs by PUID meet the need for 2 with higher granularity. I started with assuming the input for two would be file format + version as an input instead of PUIDs, but that opened up questions of formatting that seemed to be rather neatly solved by using PUIDs. Perhaps that's a decision that could use revisiting, and I'd appreciate your input as someone who's spent a lot of time around PRONOM and its data model.

Addressing 3 would I think require bringing new data into AIPscan (such as the FPR data model) and was discussed as a lower priority in the sprint meeting, so I didn't attempt to address it here.

> It seems to me like there is a more potential in this work to collapse either set of functions/reports into the other which I have described inline, i.e. the only difference between the reports seems to be
    - how the lookup happens and
    - in one report we get the PUID and in the other we don't.

This is a good point - I agree that there's room for shared code here. I take your point about keeping it DRY and combining where possible/feasible and will seek to address it in my follow-up commits. But fundamentally I do think that these reports are returning different levels of information about file formats, and I don't want to lose that distinction (in fact, it seems like maybe that distinction needs to be made more clear somehow).

> What are the implications to the user and their understanding of results when we're not distinguishing between original and preservation copy?

This is also a very good question. I think what you're pointing to here is (1) an omission from this work that I will address in follow-up commits but also (2) an opportunity for some consensus/clarity.

I think the implicit assumption in the AIPscan UI at the moment is that reports are about original files unless otherwise indicated, and to some degree this is backed up in the code (e.g. the `original_files=True` default arguments for `aip_overview` and `aip_overview_two` that we don't have flags in the GUI for). Of course, the higher level of granularity is available via the API, where intent is clearer. In the largest files report I changed this to defaulting to all files, and allowing users to limit by "original" or "preservation" file types. That was in part due to my reading of the requirements and in part anticipating that we might be keeping track of more than two types of files eventually (e.g. those Dataverse derivatives), but I'm totally fine with revisiting my assumptions and the default behaviour there.

My gut feeling for these new reports is to handle it like `aip_overview` and default to showing only original files. Perhaps this is something where a group discussion about what we want our defaults to be (and how to communicate those clearly to users) might be helpful?

> There some similarity with existing overview endpoints which I mention inline.

I do see the similarities, but I'm starting from a standpoint of disagreeing that the similarities mean that these new reports should use the existing `aip_overview` and `aip_overview_two` data endpoints.

My reasons for that are:

1. The existing endpoints (especially `aip_overview_two`, which is indeed quite similar) don't contain the data these reports require - namely, data volume per AIP of just files of a given format/PUID. So adapting that to use in this report would involves a bit more than just filtering `aip_overview_two`. Granted, we could add the data we need into that report, but...

2. From a readability and usability perspective, it seems preferable to me to have a Data endpoint where the business logic is encapsulated in a database query that gets exactly the information we want from the database without requiring post-processing and an API call that exposes that data more or less as-is. It also seems potentially less likely to result in changes in one part of the code having side effects elsewhere.

I am definitely open to further discussion. From your perspective, what makes modifying and extending the existing endpoints preferable to the approach I've taken here?

> We should try to avoid literals inline in the Python work and replace them instead with consts at an appropriate scope for both ease of changing and reuse.

I think in some cases our personal preferences on this might differ a bit (I sometimes find that at lower levels of scope adding many constants into the code can detract from readability), but I see what you're saying and will try to find a healthy balance. There might be a nice opportunity for some cleanup of existing code too, so that e.g. our request parameters are handled uniformly across the report views.

On a related note, I do think the `FIELD_` constants in `Data/data.py` could use a bit of thought/cleanup. I'm thinking specifically of moving away from camelcase as well as thinking about how the "define a const and use it everywhere" intention breaks down a bit in the templates, where we sometimes go back to hardcoding key names (e.g. to test if there's a value). I'm not sure at the moment if there's a good solution to that but will do some thinking about it.

ross-spencer commented:
> I'm going to try to start with the "meatiest" issues first and work from there .... so perhaps we can continue this conversation and add some more nuance then.

Sounds like a plan. I've a better understanding now too, so that's appreciated. 

> But fundamentally I do think that these reports are returning different levels of information about file formats, and I don't want to lose that distinction (in fact, it seems like maybe that distinction needs to be made more clear somehow).

Your description of the levels is a great help. I have no reason to disagree with the user-requirement but I think following your line of thinking around making the distinction clearer would really be beneficial.

I don't know how much I can offer that is beneficial in helping you towards a better distinction, but I might ask, is the end-use intent of any of these reports different? Focusing on originals might lend itself to preservation actions, but focusing on "any format for any file identified in the AIP" - what does that give the end-user? Does it change menu items or the UI? For example, should the reports list have sections? If we had two separate namespaces within the API, would one maybe be "preservation planning"? and the other maybe something like "file format management"? does something like that make sense? Or a literal separation "report that describes originals across AIPs", "report that describes all file formats in AIPs"? 

Just some thoughts. It sounds like your mind is already whirring so I'm looking forward to hearing more. 

The result of that thinking might lend itself to the internals of AIPscan too - so is there something in the function naming or docstrings that is missing that might have made this clearer to the reader this time around? 

And yep, maybe, let's get that group chat going later. Sometime soon. It will be easier.

> It seems to me maybe preferable in some ways to have Data endpoints where the business logic is encapsulated in a database query that gets exactly the information we want from the database without requiring post-processing (except adding context like the Storage Service name), and a corresponding API call that exposes that data more or less as-is. I find it a bit more readable (though granted that's a subjective measure), and it makes it less likely that changes in one part of the code will have side effects elsewhere.

I am really keen to hear a bit more about this. It sounds like it might be a very strong, and key, design observation/direction rather than there been a strong reason to not leverage what was there. My primary concern is if it is simply a left-turn from what was just the first approach in the first few reports, then what is the strategy to bring those first original approaches in-line and address the potential obsolescence/redundancy?

You asked should the existing endpoints be reused? I think, for me, the question is about not ignoring what is already there, and that comes from one of two perspectives: Maybe they might have been extended and reused but they haven't been in this instance... But is it an indication that at least one of the previous functions should have been rewritten, modified, or scoped for removal based on the new approach without leaving behind a partial duplication in effort/something more obsolete than not? It sounds like maybe (with the implication of extra effort) that might have been in-scope here? - there might be a communication step we need and this CR might be what that is. There might be something missing in the description of my intent around the original functions too. 

> On a related note, I do think the FIELD_ constants in Data/data.py could use a bit of thought/cleanup. I'm thinking specifically of moving away from CamelCase, as well as thinking about how the "define a const and use it everywhere" intention breaks down a bit in the templates, where we sometimes go back to hardcoding key names (e.g. to test if there's a value). I'm not sure at the moment if there's a good solution to that but will do some reading and thinking about it.

First, there was definitely thought involved :wink: it comes from my use of Golang and its struct-naming where "exported fields", which are capitalized, can be consumed automatically by the JSON library (the inverse is described nicely in the docs): _"The json package only accesses the exported fields of struct types (those that begin with an uppercase letter). Therefore only the the exported fields of a struct will be present in the JSON output."_ see: https://blog.golang.org/json though there are strategies for doing it other ways described in the decoding section of that doc I just always said to myself next time I worked on an API I'd design it at least with a default that works Golang. But then this is a team effort so we should find the best approach and a good rationale for each of our decisions. 

> "define a const and use it everywhere" intention breaks down a bit in the templates,

Totally fair. 

> I think in some cases our personal preferences on when to use constants might differ a bit (e.g. I probably wouldn't make one for an empty string, as in your inline example), 

Equally fair. That particular instance especially.

tw4l commented:
Hey @ross-spencer. Thanks again for the engaging conversation here! Let's continue offline in our meeting today, but I just wanted to quickly address a few things:

> It sounds like it might be a very strong, and key, design observation/direction rather than there been a strong reason to not leverage what was there. My primary concern is if it is simply a left-turn from what was just the first approach in the first few reports, then what is the strategy to bring those first original approaches in-line and address the potential obsolescence/redundancy?

It's interesting, this. In my mind I was very much following the existing pattern of each report having a discrete data endpoint, which feeds data into (1) an API endpoint and (2) a report in the UI. This 1:1:1 relationship, in part set out by your addition of the Data endpoint, made it easier to jump into the project and start adding reports and I think is a good conceptual model, so my intent was to continue/extend that model in this work. I think maybe where I've added a strong opinion is in having our ORM do a lot of the work in the data endpoint with a thin wrapper around it, which I do quite like but I think conceptually for me didn't change the overall model.

>  think, for me, the question is about not ignoring what is already there, and that comes from one of two perspectives: Maybe they might have been extended and reused but they haven't been in this instance... But is it an indication that at least one of the previous functions should have been rewritten, modified, or scoped for removal based on the new approach without leaving behind a partial duplication in effort/something more obsolete than not? It sounds like maybe (with the implication of extra effort) that might have been in-scope here? - there might be a communication step we need and this CR might be what that is. There might be something missing in the description of my intent around the original functions too.

I will admit, I've found finding the right scope definition for some of the AIPscan work a bit tricky, and I think you might be picking up on some of the consequences of that. In my mind these new reports were intended to be complementary, not a replacement, for those existing reports, but your prompt to question that is a valid one that might prove fruitful for thinking about the UI/user experience.

Putting on my agility hat, this might be where some good user stories connected to the existing requirements would help us guide the direction of our work.

> First, there was definitely thought involved

Oof, poor wording on my part. I certainly don't intend to be disparaging! Please chalk that up to questionable choice wording on my part and the difficulties of conveying tone via text.

> it comes from my use of Golang and its struct-naming where "exported fields", which are capitalized, can be consumed automatically by the JSON library.

Interesting! I'd love to hear more about this when we chat. I'll admit to approaching this work with a quite narrow Pythonic focus but I appreciate the extra perspective that you bring.

tw4l commented:
Hey @ross-spencer, it looks like I won't be able to get this fully in-shape for re-review quite yet - what remains is some more thinking around file types (see last WIP commit: https://github.com/artefactual-labs/AIPscan/pull/76/commits/117e790a78b62b372f6d543f7012b54b7a7951d6) and better test coverage (partially waiting on #79).

Re:

> I don't know how much I can offer that is beneficial in helping you towards a better distinction, but I might ask, is the end-use intent of any of these reports different? Focusing on originals might lend itself to preservation actions, but focusing on "any format for any file identified in the AIP" - what does that give the end-user? Does it change menu items or the UI? For example, should the reports list have sections? If we had two separate namespaces within the API, would one maybe be "preservation planning"? and the other maybe something like "file format management"? does something like that make sense? Or a literal separation "report that describes originals across AIPs", "report that describes all file formats in AIPs"?

In what I've done so far, I've added a `file_type` parameter to `Data` and `API` similar to that for the largest files endpoint (with the idea being to keep as much flexibility as possible in the backend/API) while making changes in `Reporter` so that the "AIPs by File Format" and "AIPs by PUID" reports (and the file format names and PUIDs being passed to the reports page template) are always scoped to original files only. I think that means, for now, nearly all reports in the GUI continue to cover original files only. The one exception would be the largest files report, which is different but at least makes it clear in the print report UI what is being displayed via the select and reporter header.

I do think we'll need to rethink the reports selection/generation UI at some point moving forward (in relation to #87 and #86 too) and I've some thoughts on that front (there are a couple really good lines of thought in your questions/suggestions above, and some clearer user stories might really help guide that work). But I'd prefer to scope those bigger UI questions out of this PR if you are okay with my thinking here or we can find another solution to the immediate problem at hand.

Thanks and see you Thursday! :wave:

ross-spencer commented:
Thanks Tessa - sorry I hadn't managed to get back to this until today. It was really difficult to find the head space for it last week - it sounded like the AtoM pieces were keeping you busy so I used that as my excuse :see_no_evil: 

RE: 

> I do think we'll need to rethink the reports selection/generation UI at some point moving forward (in relation to #87 and #86 too) and I've some thoughts on that front (there are a couple really good lines of thought in your questions/suggestions above, and some clearer user stories might really help guide that work). But I'd prefer to scope those bigger UI questions out of this PR if you are okay with my thinking here or we can find another solution to the immediate problem at hand.

If you were looking for initial extra feedback, then In my head the different combinations of really basic layouts for this look as follows:

![image](https://user-images.githubusercontent.com/1880412/99274197-07bafe00-27f8-11eb-9096-d821e517cc7c.png)

They'er all quite simple in their thinking though. I prefer 1. because I don't think we're going to run into a situation of 100s of lines of reports. And I feel it's pragmatic for now. 

> But I'd prefer to scope those bigger UI questions out of this PR if you are okay with my thinking here or we can find another solution to the immediate problem at hand.

It would be really great to hear more about what you're thinking and the scope of that. My feeling is that depending on the ambition of that then we can move that to another Issue, and meetup with Peter et al. and develop some stories and wire-frames this week or next - or if you have something you're confident in presenting, we can bring it to Picturae Thursday. And then for the sake of this PR, adopt something pragmatic that doesn't necessarily put us on the new track yet, but is usable. 

I know you've mentioned parameters for reports a few times and I'm aware too it's something we haven't created a pattern for yet (handling them, passing them around etc. and then what I think you're seeing, selecting them in the UI and how that changes the context of a report). With the storage service locations selection parameters on the horizon then it does seem like we'll need to solve this soon somewhere.

tw4l commented:
Hi @ross-spencer - thanks for this! And no worries about the timing - you're right that AtoM work has kept me pretty occupied.

This note is going to be a long one, but really just because I'm trying to get some thoughts down where we can reflect and maybe come back to them. Please don't feel pressured to respond to all of it, or quickly!

> If you were looking for initial extra feedback, then In my head the different combinations of really basic layouts for this look as follows

> It would be really great to hear more about what you're thinking and the scope of that. 

These UI sketches are really nice, and a great way of communicating some of the different options. Of the three options options you've outlined here I'm most a fan of #1 - I think #2 might cause some issues or confusion since the file type parameter likely won't apply for all reports (I'm thinking here of the preservation derivatives report, and user time series), and #3 might get tricky since the "type" selection would ideally change the values available in the format/PUID dropdown next to it, which is tricky without a page refresh or a more reactive UI framework to pull on.

One thing that I think might be productive to tease out a bit further is what is meant by a "generic report". Is it one where there's no potential variability in what type of files we're reporting on? As of the time I'm writing this and not yet including the reports in this PR, we have the following reports implemented in AIPscan:

* File format count: Original file formats found in all AIPs, sorted descending by count (or charted by count and size)
* AIP contents: Original file formats per AIP, with some additional information about each AIP (name, UUID, created date, size)
* Preservation copy count: Preservation derivatives per AIP, displaying original and normalized format for each
* Largest files: Largest files across all AIPs (defaults to all files, can specify to original or preservation in report UI)

Of these it seems to me that Preservation copy count might be the only one that seems to be is a truly "generic" report, in the sense that it wouldn't make sense to specify what kind of file is being reported on. Arguably variants of the others, as well as most of the reports I see forthcoming on our backlog except for the user time series, could exist under both "Original reports" (the implicit default for each at the moment) and "Derivative" reports headings. Which may not be a problem as you say, especially in the near term, given that the absolute number of reports we're talking about is relatively low. 

But given that we seem to be moving toward a future application state where we have a handful of reports, each of which may have a variable number of available report types (e.g. print table, CSV, pie chart, scatter plot, time series) and parameters (e.g. start date, end date, file type, file format, PUID), my gut says that our ideal end state would be a more reactive UI that lists a number of reports, and then allows the user to specify further refinements after one is selected (e.g. in a modal, as an in-between step to generating a report), based on what that particular selected report supports.

I played around with one example of what this might look like a bit on my own time using a very simple and rough-around-the-edges Vue app embedded in a Flask template:

![aipscan_ui_idea](https://user-images.githubusercontent.com/6758804/99453179-8dbb6f80-28f2-11eb-9df2-6342f2bc8afc.png)

![aipscan_ui_idea_modal](https://user-images.githubusercontent.com/6758804/99453185-914ef680-28f2-11eb-8c7f-9633c0eb3d0f.png)

Demo code GitHub (again, just for demo, lots of rough edges here): https://github.com/tw4l/aipscan-reports-ui-demo

Definitely not the only way to approach it, but I think one potential solution for down the road. The main benefit being that ultimately down this path, you can have the UI be responsive to the data it's being passed - e.g. displaying different report types and parameters in the modal depending on the underlying data about each report.

> My feeling is that depending on the ambition of that then we can move that to another Issue, and meetup with Peter et al. and develop some stories and wire-frames this week or next - or if you have something you're confident in presenting, we can bring it to Picturae Thursday. And then for the sake of this PR, adopt something pragmatic that doesn't necessarily put us on the new track yet, but is usable.

My experience with Vue is that it's a joy to work with, well documented, and pretty stable (Vue 3, for instance, keeps backwards-compatibility with Vue 2 apps, which is nice and rare) but does introduce new dependencies and tooling and potentially a frontend build step (the demo doesn't need one and one of the nice things about Vue is that you can sort of sprinkle it here and there into otherwise server-rendered apps, but further complexity might warrant it at some point), so is not trivial and would need to be discussed and prioritized accordingly. Even if we found a different path forward with our current approach of server-rendered code with a bit of jQuery on top, I agree that it'd be good to invest into user stories, wireframes, and some collective brainstorming and refinement before embarking in a new direction.

So for now I think it might be good to keep that larger question conversational. Which brings us back to the immediate question of what to do at this moment for these new reports.

My instinct like yours is to keep the change in this PR relatively minimal so that we can consider the UI as its own issue properly, which still presents a few options, such as:

* Keep current implicit assumption that reports are covering original files unless otherwise stated, and add new format and version reports on original files only - e.g.:

![aipscan_fornow_1](https://user-images.githubusercontent.com/6758804/99453206-97dd6e00-28f2-11eb-8408-21542e93e20e.png)

* Keep UI as-is but add original file and preservation derivative versions explicitly via naming or descriptions - e.g.: 

![aipscan_fornow_2](https://user-images.githubusercontent.com/6758804/99453222-9e6be580-28f2-11eb-88b1-dca3be587ece.png)

* Reorganize our reports into "Original files" (making what was implicit more explict) and "Preservation files", and slot the new reports in appropriately - e.g.:

![aipscan_fornow_3](https://user-images.githubusercontent.com/6758804/99453233-a592f380-28f2-11eb-8c75-266f6f3b65aa.png)

Of these my gut is telling me that the first one may suffice for now, and then we could add a preservation file version of the report when we have a supporting user story and clearer ideas on future direction. But I could certainly be swayed.

I will keep working on this (and improving test coverage, assuming #88 is merged first), but feel free to chime in with any thoughts or initial impressions as you'd like.

ross-spencer commented:
@tw4l this is most excellent! The concept for the future effort is 👌nice! Looking forward to talk more about that soon, and hope to give that demo a whirl! 

I wouldn't worry too much about "generic" in my "wire-frames" that was padding as much as anything, and an instinct that we'd end up with more than file format reports, whatever they may be? :shrug: but yeah IDK :slightly_smiling_face: 

Out of the three main options, I'd vote for two:

> Keep UI as-is but add original file and preservation derivative versions explicitly via naming or descriptions - e.g.: ... 

i really like that it simply exposes everything thus far. And doesn't hide some of what's available. It makes it absolutely clear, and lays out what needs to be done to bring everything back down into this nice compressed (potentially Vue) package you describe which would be great. 

And I really like the additional text underneath the drop-downs that you added (and it's the only image you added that to!) 

IDK - maybe Peter can tie break? :wink: 

But this is really exceptional, thank you!

tw4l commented:
Thanks @ross-spencer, looking forward to talking more about this all! Just to answer your last question - I modified the code locally and took screenshots in Firefox :laughing: Simple but effective!

ross-spencer commented on AIPscan/templates/report_base.html:
```diff
@@ -0,0 +1,27 @@
+    <link rel="shortcut icon" href="{{ url_for('static', filename='favicon.ico') }}">
+    <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/jquery-ui.css') }}">
+    <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/bootstrap.min.css') }}">
+    <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/custom.css') }}">
```
@tw4l I was wondering about this as I work on the agents table. I used a custom CSS, say `style_agents_report.css`. This grants some flexibility around the amount of customization I can do in that table. Minor stuff, but it also leaves out the ambiguity of having to check whether a style is being used in more than one report. 

Does that make sense to you? And if so, can we split `base.html` into two? (although I am not sure how that would with the need for a `</head>` end tag for example. 

Or should I perhaps use `custom.css` and just use comments to separate sections, but accept that stylesheet might grow over time? 

It looks like I am coming to a pause on this one tonight, so will get back to it tomorrow.

tw4l replied:
Hi @ross-spencer - this is a good question!

We could use a single `custom.css` as you say, but another option might be to use an additional block in order to add custom CSS per report.

Looking at Archivematica as an example, we could do something like this in the `report_base.html` template: https://github.com/artefactual/archivematica/blob/stable/1.12.x/src/dashboard/src/templates/layout.html#L14

And then in our reports templates, add the additional CSS like this: https://github.com/artefactual/archivematica/blob/stable/1.12.x/src/dashboard/src/templates/archival_storage/archival_storage.html#L19-L24

Those are Django templates and not jinja2, of course, but I think it should work the same way.

tw4l replied:
Doing the same for the jQuery/JS sprinkled into the templates might be a nice way to cleanly separate that out as well.

tw4l commented on AIPscan/Reporter/templates/reports.html:
```diff
@@ -154,16 +149,26 @@
+{% else %}
+
+    <div style="padding:1em;">
+        Storage Service information is not yet available for reporting. <a href="{{ url_for('aggregator.storage_services') }}">Configure a Storage Service</a> and run a Fetch Job to to get started.</a>
```
typo: "to to"

tw4l commented:
Hey @ross-spencer - long time in the coming but this is now ready for re-review. I think I've addressed all of your comments and questions but please let me know if there's something I missed.

On the reports UI, I did end up implementing something close to my second option above for now:

![aipscan_reports_with_puids](https://user-images.githubusercontent.com/6758804/100145798-daf58f00-2e66-11eb-88ac-baafadc5847e.png)

In many cases we don't have PUIDs for preservation derivatives (they seem to be present in the METS only when people input them during manual normalization workflows?) so for many users that screen will likely look more like this:

![aipscan_reports_no_puids](https://user-images.githubusercontent.com/6758804/100145838-f19be600-2e66-11eb-8893-81793ee63948.png)

I also added some helpful messages to display in place of the reports when there are no Storage Services configured, or a Fetch Job hasn't been run yet for the selected Storage Service.

Other changes include:
* Reduced duplication between `aips_by_file_format` and `aips_by_puid` Data endpoints
* Removing error handling for missing parameters in the API endpoints (they now return an empty list instead)
* Adding an original_files parameter (with default True) for the reports
* Improved test coverage, using our new ability to set test state in a fixture (and adding one new such fixture to the Data blueprint's tests module)
* Better `_get_storage_service` logic in the Data and Reporter blueprints (I noticed that the previous logic for the former made it possible for the endpoints to return false/inconsistent data, so patched that up while I was there)
* Better use of constants, and in the case of request parameters, a new `request_params` shared dict
* Better exception handling for `AttributeErrrors` when attempting to access the new `StorageService` model properties in the reports view.
* Wrapping `natsort` in a more descriptive helper function for naturally sorting PUIDs.
* Making the `namespace_data` API routes consistently semantic - e.g. `/storage_service/<storage_service_id>/file-format` rather than `file-format/<storage_service_id>`

Not included in this piece of work is non-tabular reports and start/end data parameters, which as noted before I've intentionally left out for now and will address in a separate PR so they don't hold up this work.

Thanks!

ross-spencer commented:
Awesome @tw4l I anticipate reviewing this today - in-fact - I plan to not leave my desk until review is done!

ross-spencer commented on AIPscan/Reporter/views.py:
```diff
@@ -133,16 +136,30 @@ def view_file(file_id):
+    if storage_service_id is not None:
+        storage_service = StorageService.query.get(storage_service_id)
+    else:
+        storage_service = StorageService.query.filter_by(default=True).first()
```
This is looking good! Nice work!

ross-spencer reviewed (commented):
Thanks @tw4l this is looking really great. Perhaps it will be prudent to check what has happened to the alert around the yet to be implemented `Format version count` report. Perhaps that alert code can be removed entirely now the page isn't so bear?

Otherwise I have left some comments that might also be worth getting your opinion on. Possibly the only one that might result in a change is around the way file formats are grouped for searching on. This is very much related to PRONOM's imperfections and I personally would love to figure out with y'all how to make the UI as clear as possible about what these names mean, i.e. they're not quite distinct file formats, and they're not quite format families either. It's just a grouping by name, and might end up being a mix of those two things.

Related, I think there might be one issue worth logging as a follow-up about an endpoint for those format names. So a workflow for a user of the API might be:

1. Query format name groups. 
2. Query for the report-like statistics based on that filter. 

Around the storage service ID/endpoints I have suggested an issue be logged about discovering storage service resource information. Specifically the ID. The intention of the previous code was to always return something, and I agree the result might be more unintuitive than perhaps unpredictable, but not necessarily desirable. We should be able to ask for correct information and get correct information and likewise ask for bad information and get nothing back as it's a bit "undefined".

Just some miscellaneous observations:

* It's an important PR with some good examples demonstrating how to stabilize this work.
* Docstrings are looking good. I don't know if you've seen my read the docs PR but they'll lend themselves well to that.
* Some really good examples of "clean code" here.

tw4l commented on AIPscan/Data/data.py:
```diff
@@ -47,11 +49,17 @@
+
+    Unlike elsewhere in our application, here we do not fall back to
+    a different StorageService if the user-supplied ID is invalid to
+    prevent inaccurate information from being returned.
```
I think that's a great suggestion. I'll log one or more issues for API endpoints to add - namely, this as well as listing available format names and PUIDs.

tw4l commented on AIPscan/Data/data.py:
```diff
@@ -2,8 +2,10 @@
+from AIPscan import db
 from AIPscan.models import AIP, File, FileType, StorageService
 
+VALID_FILE_TYPES = set(item.value for item in FileType)
```
I've been finding using the `FileType` enum to be pretty smooth. I'm not quite sure I follow your suggestion around disposition - do you mean to change the name `file_type`/`FileType` or are you envisioning a larger change?

ross-spencer replied:
Indeed! A name change - mainly because file type means a lot more things to me (especially if we're in the land of formats), and not specifically "'original" or "preservation". but YMMV.

tw4l replied:
Hm, yeah that's fair! I'm not married to "file type" but do think maybe that'd be better changed in a separate PR and all at once?

ross-spencer replied:
Agreed!

tw4l commented on AIPscan/Data/data.py:
```diff
@@ -292,3 +299,130 @@ def largest_files(storage_service_id, file_type=None, limit=20):
+
+    storage_service = _get_storage_service(storage_service_id)
+    report[FIELD_STORAGE_NAME] = ""
+    try:
+        report[FIELD_STORAGE_NAME] = storage_service.name
+    except AttributeError:
+        pass
```
You know, you make a good point. I can't remember why I put in this try/except block now but you're right that it doesn't seem likely, so I will change this.

The question about using the ID rather than the name is interesting. I'm tempted to stick with the name just because the Storage Service ID is rather internal to AIPscan. Another thing we could consider using moving forward is the Storage Service URL, but I'm not sure if including that in reports would have security implications.

If it's alright with you, I will revert this to `report[FIELD_STORAGE_NAME] = storage_service.name` for this PR, and maybe we can keep pondering how to best represent Storage Service information in the reports separately as the need arises.

ross-spencer replied:
Sounds good to me, and great points about that.

tw4l commented on AIPscan/Reporter/templates/reports.html:
```diff
@@ -48,33 +47,128 @@
+            <td>
+                File format
+                <br>
+                <span class="text-muted">AIPs containing original files in given file format.</span>
```
I completely understand what you're getting at but I wonder if end users would be familiar with the term "format naming group", or if it might have the potential to cause confusion.

I've tried to be consistent with the FPR data model (or at least the `Format` and `FormatVersion` subset) as much as possible in this work when considering what a "format" vs. "format version" is, with the aim of at least being consistent within the Archivematica ecosystem, but I agree there are issues here, as the PDF example makes very clear. From my understanding, Archivematica/the FPR always makes identifications at the `FormatVersion` level (and that's where we find PUIDs as well, which is why I thought it would be safe to use the PUID as a proxy for file format version), and every `FormatVersion` belongs to a `Format` (as a many-to-one relationship, but in practice sometimes it's one-to-one in our data) so I've tried to stick to those terms as much as possible, even when they're not quite how I'd use them outside of this context.

That said, this is a tricky one and I don't feel very confident either way so if you have a strong leaning here I'd be happy to make the change and maybe we can try to get some user validation/feedback down the road?

ross-spencer replied:
Definitely tricky but I can understand your approach. 

I'm reminding myself of trivia when someone thinks something might be the right answer, but doesn't have an answer :smile: 

I suspect this might get clearer/easier with a PRONOM-like (ish) integration in future so that's in the back of my mind. And some validation/feedback will help - if it works it works. 

My leaning is to try and do something hearing what you say about format naming group, I think maybe that won't work then. But at the same time, I'd love to not assert this aggregation as a format version. 

With your not to alignment with the FPR what about adding the PUID IDs in brackets alongside the title? 

![image](https://user-images.githubusercontent.com/1880412/100390461-6fdec080-2ffe-11eb-8fd7-77b789f35896.png)

Does that work? Maybe we can each chat offline too if it doesn't quite hit the mark.

tw4l replied:
We discussed a bit offline and, although I like the idea, adding the PUIDs in brackets might introduce some new UI issues and might be tricky with the data that we currently have in AIPscan, so we will leave this as-is for now and I will file a new issue describing the problem and some of our thinking to date on it.

tw4l commented on AIPscan/API/namespace_data.py:
```diff
@@ -22,19 +29,12 @@
-
-
-@api.route("/aip-overview/<storage_service_id>")
+@api.route("/storage_service/<storage_service_id>/aip-overview")
```
Thanks for this Ross - this is a really nice and helpful backstory/rationale and I appreciate the detail! Your thinking makes sense to me too.

Let me start by saying that I admit I may have overreached here, so very possibly the best thing to do is to simply revert the routes to how they were before.

My motivation for this commit was always inspired a bit by the idea of cool URIs, and particularly how I typically expect the components of the URI to be in a semantically meaningful order, especially when some sort of identifier is being used.

So looking at the `/api/ingest/<SIP UUID>/delete/` Archivematica API endpoint as an example, we might see something in the wild like: `https://example.com/api/ingest/9259c0bb-5636-4420-a6f3-ee283b32d5a4/delete`. The UUID here relates for the semantic unit that immediately precedes it (in this case "ingest"), and the next component of the URI (in this case "delete"), is an action taken on the resource previously identified.

An existing AIPscan API endpoint `fmt-overview/<storage_service_id>` might look in the wild like this: `https://example.com/api/fmt-overview/1`. I found it odd that the ID here is identifying a resource type that's not found in the URI, hence the reordering.

> If I /try/ to read your intention - I wonder if perhaps between the two positions we'll find that storage service ID might be better as a query parameter? IDK.

You know, I think you might be right here! But maybe again these are higher-level questions that would benefit from a more direct focus on the API rather than done as a "by the way" commit in a PR like this, so I will revert them for now.

tw4l replied:
And on a higher-level note, I'm recognizing here that on my end there are probably two issues with how I approached this: 1. trying to cram it in a only-sort-of-related PR, and 2. not communicating my intentions clearly enough, either in code, an issue, or elsewhere. So room for improvement on my end, which is noted and I will try to bear in mind moving forward!

tw4l commented on AIPscan/Data/data.py:
```diff
@@ -292,3 +299,130 @@ def largest_files(storage_service_id, file_type=None, limit=20):
+
+    :param storage_service_id: Storage Service ID (int)
+    :param search_string: File format or PUID (str)
+    :param original_files: Boolean indicating whether returned data
+    describes original (default) or preservation files.
```
Yes, that's nice!

tw4l commented:
@ross-spencer Thank you for the review and the encouragement! This has been a long and involved PR for both of us, but in the end what's getting merged in is much improved by your careful attention, questions, and insights :smile: 

I believe that the remaining issues we've agreed to change in this PR have now been fixed in the previous four commits.

In addition, I've created several new issues based on your comments and our discussions in this PR:

- #94 : AIPscan is missing an API endpoint for listing Storage Service locations
- #95: AIPscan is missing API endpoints for listing the file formats and PUIDs in a Storage Service
- #96: The term "file format" as it's used in AIPscan might be a bit misleading (I did have some difficulty writing this one and feel like there's likely a simpler way to describe it, so please feel free to edit/add!)
- #97: New report parameters may require a rethinking of the reports page UI

ross-spencer reviewed (approved):
Superb, thanks Tessa, and likewise, I appreciate your patience, communication and perseverance! This is looking great!

One last thing to say... 

LGTM! 

![LGTM](https://media4.giphy.com/media/3o6UB0WHErPubNDTfa/giphy.gif?cid=ecf05e477qs1wu4ke9n0wbylrfbbihtatg8on7w3e9y935v3&rid=giphy.gif)
